File: 151004/Glushachenko/lab3/cassandra/Publisher/Service/Implementation/AuthorService.cs
using AutoMapper;
using Publisher.Entity.Db;
using Publisher.Entity.DTO.RequestTO;
using Publisher.Entity.DTO.ResponseTO;
using Publisher.Repository.Interface;
using Publisher.Service.Implementation.Common;
using Publisher.Service.Interface;

namespace Publisher.Service.Implementation
{
    public class AuthorService(IMapper mapper, IAuthorRepository repository) :
        AbstractCrudService<Author, AuthorRequestTO, AuthorResponseTO>(mapper, repository), IAuthorService
    {
        private readonly IMapper _mapper = mapper;

        public override async Task<AuthorResponseTO> Add(AuthorRequestTO authorTo)
        {
            if (!Validate(authorTo))
            {
                throw new InvalidDataException("Author is not valid");
            }

            return await base.Add(authorTo);
        }

        public override async Task<AuthorResponseTO> Update(AuthorRequestTO authorTo)
        {
            if (!Validate(authorTo))
            {
                throw new InvalidDataException($"UPDATE invalid data: {authorTo}");
            }

            return await base.Update(authorTo);
        }

        public async Task<AuthorResponseTO> GetByTweetID(int tweetId)
        {
            var response = await repository.GetByTweetIdAsync(tweetId);

            return _mapper.Map<AuthorResponseTO>(response.Author);
        }

        private static bool Validate(AuthorRequestTO author)
        {
            var fnameLen = author.FirstName.Length;
            var lnameLen = author.LastName.Length;
            var passLen = author.Password.Length;
            var loginLen = author.Login.Length;

            if (fnameLen < 2 || fnameLen > 64)
            {
                return false;
            }
            if (lnameLen < 2 || fnameLen > 64)
            {
                return false;
            }
            if (passLen < 8 || passLen > 128)
            {
                return false;
            }
            if (loginLen < 2 || loginLen > 64)
            {
                return false;
            }
            return true;
        }
    }
}

File: 151004/Afanasenko/appLab5/common/src/main/kotlin/com/github/kornet_by/dc/lab5/service/AuthorService.kt
package com.github.kornet_by.dc.lab5.service

import com.github.kornet_by.dc.lab5.dto.request.AuthorRequestTo
import com.github.kornet_by.dc.lab5.dto.request.AuthorRequestToId
import com.github.kornet_by.dc.lab5.dto.response.AuthorResponseTo

interface AuthorService {
	suspend fun create(requestTo: AuthorRequestTo?): AuthorResponseTo?

	suspend fun deleteById(id: Long): Boolean

	suspend fun getAll(): List<AuthorResponseTo>

	suspend fun getById(id: Long): AuthorResponseTo?

	suspend fun update(requestTo: AuthorRequestToId?): AuthorResponseTo?
}